.docker/Dockerfile

```
@@ -0,0 +1,33 @@
+# start with official ruby docker image as base
+FROM ruby:3.1.2
+
+# set working directory within container
+WORKDIR /usr/src/app
+
+# pull in ruby (jekyll) and python (cite process) package info
+COPY Gemfile Gemfile.lock ./_cite/requirements.txt .
+
+# command to get bundler version
+ARG BUNDLER="grep -A 1 'BUNDLED WITH' Gemfile.lock | tail -n 1 | xargs"
+
+# install ruby bundler
+RUN gem install bundler --version $(eval ${BUNDLER})
+
+# install ruby packages
+RUN bundle _$(eval ${BUNDLER})_ install
```

Why did you put code in this build arg? Not only is this uncommon, but if you were to attempt to specify the value of the build arg via, e.g., `--build-arg BUNDLER=2.4.7` in the build step, the lines where it's evaluated would fail since `eval "2.4.7"` will raise an error.

I've tried to get it to work the way that I think you intended, i.e. if the `BUNDLER` build arg isn't specified, it should run some code to evaluate its value from `Gemfile.lock` in the Dockerfile itself. I've run into a few issues, none of which are insurmountable, but IMHO make the file confusing for someone familiar with Docker. One, variables, even exported ones, don't survive from one `RUN` directive to another, since each is evaluated in a separate shell. This either means that we'd need either need to redundantly resolve it in each `RUN` directive, or combine all the`RUN` directives that use it into a single directive. Also, since BUNDLER is a build arg, we can't simply overwrite $BUNDLER if it's not defined; it'll be replaced with its value by Docker in each directive before any code is run, reverting the attempt to set it; you'd need another variable, which IMHO adds to the noise in this file.

Anyway, the "tl;dr" is that I'd suggest you resolve the bundler version in `run.sh` and pass it as a build arg; that would IMHO be the method most familiar to someone who uses Docker.

---

This is a workaround provided in this thread: https://github.com/moby/moby/issues/29110 It is just a method to assign a variable to keep things DRY. The intent was to not have bundler version as a build arg at all. To me having it as a build arg doesn't make sense, as the container will have the Gemfile.lock file directly, it doesn't co-locate concerns, and it makes the caller do extra unnecessary work (might be important if we pre-build and publish the image).

If you think the ARG method is too hacky, I will move it back to the run script.

---

Ah, sorry, I threw in a build arg in the original without asking you if that's something you wanted.

In the case that you don't anticipate people to want to override the bundler version (say, if they built the image themselves, which is what build args are for), I'd just the bundler install and gem installation in a single RUN to keep things DRY. I've tested that this works:

```suggestion
RUN BUNDLER_VER=$( grep -A 1 'BUNDLED WITH' Gemfile.lock | tail -n 1 | xargs ) ; \
    gem install bundler --version ${BUNDLER_VER} && \
    bundle _${BUNDLER_VER}_ install
```

---

Right, sorry, what I meant was giving the person building the image a build arg if they wanted to override the part where it reads from `Gemfile.lock`, but thinking on it I don't anticipate anyone will want to do that. The process supports adding gems just fine, as the `Gemfile` and `Gemfile.lock` files are copied in and used to install everything.

Frankly, I think it's fine not to verify the lockfile, as you're using the `Gemfile` as the source of truth; you might even consider skipping and/or ignoring the lockfile entirely, aside from reading the bundler version from it. Lockfiles are typically used when you need other deployments to use the exact versions of every package that you had installed, which includes platform specifics, as you observed. In this case, since you want users to edit the `Gemfile`, they'll end up with the gem versions as specified from the `Gemfile`, not from the lockfile, anyway.

I have to admit I didn't think hard about the lines I'd copied from a Dockerfile for building a GitHub pages site, so I'm fine with you removing things that don't make sense or that you can't vouch for, because I can't vouch for them, either.

---

> Lockfiles are typically used when you need other deployments to use the exact versions of every package that you had installed

Indeed, but Ruby doesn't seem to work the same way as any other language with the lock file. You edit it to add platforms, it specifies a bundler version, it changes depending on what platform you build it on... It's all very unintuitive to me and I don't understand how it works, so I'm hesitant to remove the lockfile.

The *real* solution to all of this is to not use Jekyll/Ruby at all (or ever again, for my taste), and go to something node-based so it's all in the same languages (markdown/html/css/js), and reliably cross platform with a single-install (I've never had platform compatibility issues with node).

---

References I'll read through:
https://stackoverflow.com/questions/4151495/should-gemfile-lock-be-included-in-gitignore
https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/

Some of these comments seem to indicate that if you're working on something that can be built on multiple platforms, you should actually gitignore Gemfile.lock... maybe I will experiment with this.

---

FWIW I wasn't recommending removing the lockfile from the repo, just from the docker image; it's a useful artifact to have even if it's not relied upon in your build process.

I don't know what other languages you've seen with lockfiles, but ones that build packages from source (which Ruby does) need to use a version that's compatible with the target platform, so platform-specific information is typically included. JavaScript is fully interpreted and the developer-side stuff targets node, which is effectively a VM that was carefully written for each platform it targets, so I'd consider it a special case for lockfiles, not the norm. (Anecdotally, I've had issues with node on the rare occasions that you do need a platform-specific binary built from source. IIRC I was depending on some less/sass preprocessor that built a binary and every time my dev environment would change it would break. Anyway, it's been a long time since then and it's probably been fixed to use a pure JS implementation, as I remember it causing everyone pain...)

Also, sure, I don't love Ruby (and as a result, Jekyll) either, and I'd be happy to see you replace it with node.